auth_token/models.py

```diff
@@ -129,55 +171,357 @@ def delete(self):
+        qs = self.filter(
+            is_active=True,
+            slug=slug,
+        )
+
+        if isinstance(obj_or_class, models.Model):
+            qs = qs.filter(object_id=obj_or_class.pk)
+        return qs.filter(key=key) if key else qs
+
+
+class OneTimePasswordManager(SmartManager):
+
+    def create(self, key_generator, **kwargs):
+        for attempt in range(settings.MAX_RANDOM_KEY_ITERATIONS + 1):
+            try:
+                key = key_generator()
+                hashed_key = hash_key(key)
+                obj = super().create(key=hashed_key, **kwargs)
+                obj.secret_key = key
+                return obj
+            except IntegrityError:
+                if attempt > settings.MAX_RANDOM_KEY_ITERATIONS:
+                    raise IntegrityError('Could not produce unique key for authorization token')
+
+
+class OneTimePassword(SmartModel):
+    """
+    Specific verification tokens that can be send via e-mail, SMS or another transmission medium
+    to check user authorization (example password reset)
+    """
+    key = models.CharField(
+        verbose_name=_('key'),
+        max_length=128,
+        primary_key=True,
+        null=False,
+        blank=False
+    )
+    expires_at = models.DateTimeField(
+        verbose_name=_('expires at'),
+        null=True,
+        blank=True,
+    )
+    slug = models.SlugField(
+        verbose_name=_('slug'),
+        null=False,
+        blank=False
+    )
+    is_active = models.BooleanField(
+        verbose_name=_('is active'),
+        default=True
+    )
+    data = models.JSONField(
+        verbose_name=_('data'),
+        null=True,
+        blank=True,
+        encoder=DjangoJSONEncoder
+    )
+    related_objects = GenericManyToManyField()
+
+    secret_key = None
+
+    objects = OneTimePasswordManager()
+
+    class Meta:
+        ordering = ('-created_at',)
+        verbose_name = _('one time password')
+        verbose_name_plural = _('one time passwords')
+
+    def check_key(self, key):
+        return self.key == hash_key(key)
+
+    @property
+    def is_expired(self):
+        return self.expires_at and self.expires_at < timezone.now()
+
+
+class AuthorizationRequest(SmartModel):
+
+    authorization_token = models.ForeignKey(
+        verbose_name=_('authorization token'),
+        to=AuthorizationToken,
+        related_name='authorization_requests',
+        null=True,
+        blank=True,
+        on_delete=models.SET_NULL
+    )
+    user = models.ForeignKey(
+        verbose_name=_('user'),
+        to=django_settings.AUTH_USER_MODEL,
+        related_name='authorization_requests',
+        null=False,
+        blank=False,
+        on_delete=models.CASCADE
+    )
+    mobile_device = models.ForeignKey(
+        verbose_name=_('mobile device'),
+        to=MobileDevice,
+        related_name='authorization_requests',
+        null=True,
+        blank=True,
+        on_delete=models.CASCADE
+    )
```

Comment:
A lot of foreign keys, some of them can be null, wouldn't be better to save them as related objects?

Reply:
I was thinking about this, but the searching will be complicated. This should not be generic relation